oioswift/common/middleware/slo.py

```diff
@@ -0,0 +1,349 @@
+                )
+            seg_data = {
+                'name': '/' + seg_dict['path'].lstrip('/'),
+                'bytes': seg_resp['bytes'],
+                'hash': seg_resp['hash'],
+                'content_type': seg_resp['content_type'],
+                'last_modified': last_modified_formatted
+            }
+            if seg_dict.get('range'):
+                seg_data['range'] = seg_dict['range']
+            if config_true_value(seg_resp['slo']):
+                seg_data['sub_slo'] = True
+
+            return segment_length, seg_data
+        # END: Adapt for OpenIO code
+
+        heartbeat = config_true_value(req.params.get('heartbeat'))
+        separator = ''
+        if heartbeat:
+            # Apparently some ways of deploying require that this to happens
+            # *before* the return? Not sure why.
+            req.environ['eventlet.minimum_write_chunk_size'] = 0
+            start_response('202 Accepted', [  # NB: not 201 !
+                ('Content-Type', out_content_type),
+            ])
+            separator = '\r\n\r\n'
+
+        def resp_iter(total_size=total_size):
+            # wsgi won't propagate start_response calls until some data has
+            # been yielded so make sure first heartbeat is sent immediately
+            if heartbeat:
+                yield ' '
+            last_yield_time = time.time()
+
+            # BEGIN: New OpenIO code
+            sub_req = make_subrequest(
+                req.environ,
+                path='%s?format=json&prefix=%s&limit=%d' %
+                     (segments_container_path, seg_prefix,
+                      self.max_manifest_segments),
+                method='GET',
+                headers={'x-auth-token': req.headers.get('x-auth-token')},
+                agent='%(orig)s SLO MultipartPUT', swift_source='SLO')
+            sub_req.environ.setdefault('oio.query', {})
+            # All meta2 databases may not be synchronized
+            sub_req.environ['oio.query']['force_master'] = True
+            sub_req.environ['oio.query']['slo'] = True
+            list_seg_resp = sub_req.get_response(self)
+
+            with closing_if_possible(list_seg_resp.app_iter):
+                segments_resp = json.loads(list_seg_resp.body)
+
+            seg_resp_dict = dict()
+            for seg_resp in segments_resp:
+                obj_name = '/'.join(('', segments_container, seg_resp['name']))
+                seg_resp_dict[obj_name] = seg_resp
+
+            for obj_name in path2indices:
+                now = time.time()
+                if heartbeat and (now - last_yield_time >
+                                  self.yield_frequency):
+                    # Make sure we've called start_response before
+                    # sending data
+                    yield ' '
+                    last_yield_time = now
+
+                for i in path2indices[obj_name]:
+                    seg_dict = parsed_data[i]
+                    obj_name = seg_dict['path']
```

Comment:
This statement is either useless or dangerous: `obj_name` is the external loop variable.